### PR TITLE
Issue 1823/ingredients taxonomy

### PR DESCRIPTION
**Pull request by stephanegigandet**
Combine ingredients, additives, vitamins etc. in one single taxonomy.

Add some checks for taxonomy definition, + changes to the taxonomies to make the test pass.

### CODE REVIEW

**Comment by teolemon** [comment]
``` 
#   Failed test at t/tags.t line 294.
#     Structures begin differing at:
#          $got->[1] = 'en:fruit-juice'
#     $expected->[1] = 'en:citrus-fruit'
# [
#   'en:fruit',
#   'en:fruit-juice',
#   'en:citrus-fruit',
#   'en:salt',
#   'en:sugar',
#   'en:orange',
#   'en:orange-juice',
#   'en:concentrated-orange-juice'
# ]
#   Failed test at t/tags.t line 309.
#     Structures begin differing at:
#          $got->[2] = 'en:fruit-juice'
#     $expected->[2] = 'en:citrus-fruit'
# [
#   'en:concentrated-orange-juice',
#   'en:fruit',
#   'en:fruit-juice',
#   'en:citrus-fruit',
#   'en:orange',
#   'en:orange-juice',
#   'en:sugar',
#   'en:salt'
# ]
# Looks like you failed 2 tests of 33.
t/tags.t ......................... 
Dubious, test returned 2 (wstat 512, 0x200)
Failed 2/33 subtests 
t/text.t ......................... ok
```

**Comment by aleene** [comment]
Are there now extra requirements for ingredients with multiple parents?

**Comment by stephanegigandet** [comment]
@aleene : no extra requirements for ingredients with multiple parents. There are now tests to make sure that there aren't entries or synonyms that belong to 2 different entries. (although for ingredients, the taxonomy will build even if we have an entry multiple times, at least for now)

File: scripts/build_tags_taxonomy.pl [hunk 1]
```diff
@@ -25,14 +25,51 @@
...
+
+my $file = $tagtype . ".txt";
+
+# For the Open Food Facts ingredients taxonomy, concatenate additives, minerals, vitamins, nucleotides and other nutritional substances taxonomies
+
+if (($tagtype eq "ingredients") and ($server_domain =~ /openfoodfacts/)) {
+
+	$file = "ingredients.all.txt";
+		
+	open (my $OUT, ">:encoding(UTF-8)", "$data_root/taxonomies/$file") or die("Cannot write $data_root/taxonomies/$file : $!\n");
```
Comment (hangy) [inline scripts/build_tags_taxonomy.pl]: Does the taxonomy parser work with synonyms and stopwords for a language potentially being defined several times?

File: t/tags.t [hunk 2]
```diff
@@ -293,24 +293,30 @@ my @tags = ();
...
    'en:orange-juice',
    'en:concentrated-orange-juice'
  ]
- ) or diag explain(\@tags);;
+ ) or diag explain(\@tags);
+
+
+foreach my $tag (@tags) {
+
+	print STDERR "tag: $tag\tlevel: " . $level{ingredients}{$tag} . "\n";
```
[the remark this answers is not in the document]
Comment (stephanegigandet) [reply t/tags.t]: We keep having this test fail, that's why I added some debug info. We probably should relax this test in some way, or find a way to run it on a static taxonomy.

File: scripts/build_tags_taxonomy.pl (hunk 1 above)
Comment (stephanegigandet) [reply scripts/build_tags_taxonomy.pl]: It should work.

**Pull request merged**